src/app/qgsmaptoolreverseline.cpp

```diff
@@ -0,0 +1,190 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsmaptoolreverseline.h"
+
+#include "qgsfeatureiterator.h"
+#include "qgsmapcanvas.h"
+#include "qgsvertexmarker.h"
+#include "qgsvectorlayer.h"
+#include "qgsgeometry.h"
+#include "qgsrubberband.h"
+#include "qgssnappingutils.h"
+#include "qgstolerance.h"
+#include "qgisapp.h"
+#include "qgslinestring.h"
+#include "qgsmultilinestring.h"
+
+#include <QMouseEvent>
+
+QgsMapToolReverseLine::QgsMapToolReverseLine( QgsMapCanvas *canvas )
+  : QgsMapToolEdit( canvas )
+  , mPressedFid( 0 )
+  , mPressedPartNum( 0 )
+{
+  mToolName = tr( "Reverse line geometry" );
+}
+
+QgsMapToolReverseLine::~QgsMapToolReverseLine()
+{
+  delete mRubberBand;
+}
+
+void QgsMapToolReverseLine::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  Q_UNUSED( e );
+  //nothing to do
+}
+
+void QgsMapToolReverseLine::canvasPressEvent( QgsMapMouseEvent *e )
+{
+  mPressedFid = -1;
+  mPressedPartNum = -1;
+  delete mRubberBand;
+  mRubberBand = nullptr;
+
+  QgsMapLayer *currentLayer = mCanvas->currentLayer();
+  if ( !currentLayer )
+    return;
+
+  vlayer = qobject_cast<QgsVectorLayer *>( currentLayer );
+  if ( !vlayer )
+  {
+    notifyNotVectorLayer();
+    return;
+  }
+
+  if ( !vlayer->isEditable() )
+  {
+    notifyNotEditableLayer();
+    return;
+  }
+
+  QgsGeometry geomPart = partUnderPoint( e->pos(), mPressedFid, mPressedPartNum );
+
+  if ( mPressedFid != -1 )
+  {
+    mRubberBand = createRubberBand( vlayer->geometryType() );
+
+    mRubberBand->setToGeometry( geomPart, vlayer );
+    mRubberBand->show();
+  }
+
+}
+
+void QgsMapToolReverseLine::canvasReleaseEvent( QgsMapMouseEvent *e )
+{
+  Q_UNUSED( e );
+
+  delete mRubberBand;
+  mRubberBand = nullptr;
+
+  if ( !vlayer || !vlayer->isEditable() )
+  {
+    return;
+  }
+
+  if ( mPressedFid == -1 )
+    return;
+
+  QgsFeature f;
+  vlayer->getFeatures( QgsFeatureRequest().setFilterFid( mPressedFid ) ).nextFeature( f );
+  QgsGeometry geom;
+
+  if ( f.geometry().constGet() )
+  {
+    if ( f.geometry().isMultipart() )
+    {
+      std::unique_ptr<QgsMultiCurve> line_reversed( static_cast<QgsMultiCurve * >( f.geometry().constGet()->clone() ) );
+      std::unique_ptr<QgsCurve> line_part( static_cast<QgsCurve *>( line_reversed->geometryN( mPressedPartNum )->clone() ) );
+      std::unique_ptr<QgsCurve> line_part_reversed( line_part->reversed() );
+      line_reversed->removeGeometry( mPressedPartNum );
+      line_reversed->insertGeometry( line_part_reversed.release(), mPressedPartNum );
+
+      geom = QgsGeometry( line_reversed.release() );
+
+    }
+    else
+    {
+
+      std::unique_ptr<QgsCurve> line( static_cast<QgsCurve *>( f.geometry().constGet()->clone() ) );
```

Comment:
There's an extra clone happening here which isn't needed - you should be able to get away with

    geom = QgsGeometry( static_cast< const QgsCurve* >( f.geometry().constGet() )->reversed() );